sources/declarative/decodable/DecoderSequence.swift

```diff
@@ -0,0 +1,31 @@
+open class CSVDecoderSequence: IteratorProtocol, Sequence {
+    private let source: ShadowDecoder.Source
+    private var currentIndex: Int = 0
+
+    init(source: ShadowDecoder.Source) {
+        self.source = source
+    }
+
+    /// Advances to the next row and returns a `CSVRowDecoder`, or `nil` if no next row exists.
+    public func next() -> CSVRowDecoder? {
+        guard !self.source.isRowAtEnd(index: self.currentIndex) else {
+            return nil
+        }
+
+        defer { self.currentIndex += 1 }
+        let decoder = ShadowDecoder(source: source, codingPath: [IndexKey(self.currentIndex)])
+        return CSVRowDecoder(decoder: decoder)
+    }
+}
+
+public struct CSVRowDecoder {
+    /// The representation of the decoding process point-in-time.
+    let decoder: ShadowDecoder
```

Comment:
Let's make it `private` and then have a `fileprivate` initializer to pass the `decoder`.